### PR TITLE
Auto reauth; Closes #152

### LINKED ISSUE #152: Handle automatic re-auth

### PR DESCRIPTION
**Pull request by jrperritt**

### CODE REVIEW

**Comment by jrperritt** [comment]
Originally I added a `MaxRetries` field to the `AuthOptions` structure to make the number of re-authentication retries variable, but I don't know how helpful that'd be. If there is interest in adding it, it would be just a few lines. Currently it just tries to reauth once.

File: openstack/client.go [hunk 1]
```diff
@@ -107,6 +108,10 @@ func v2auth(client *gophercloud.ProviderClient, endpoint string, options gopherc
 		return err
 	}
 
+	if options.AllowReauth {
+		client.ReauthFunc = AuthenticateV2
+		client.AuthOptions = options
+	}
```
Comment (smashwilson) [inline openstack/client.go]: You could also capture `AuthOptions` (and the client pointer) with a closure, instead:

```go
client.ReauthFunc = func() error {
  return AuthenticateV2(client, options)
}
```

Pros:
 * Cleaner (in my opinion, anyway).
 * Avoids unnecessarily exporting `AuthOptions` in the client struct.
 * Can capture more context in the future if we ever need it without breaking the function's type signature and therefore backwards compatibility (although we'd most likely need to add it to `AuthOptions` instead).

Cons:
 * A little more magical

File: provider_client.go [hunk 2]
```diff
@@ -179,6 +189,19 @@ func (client *ProviderClient) Request(method, url string, options RequestOpts) (
 		return nil, err
 	}
 
+	if resp.StatusCode == 401 {
```
Comment (smashwilson) [inline provider_client.go]: You could also use the constant here to avoid the magic number; `http.StatusUnauthorized`.

File: rackspace/client.go [hunk 3]
```diff
@@ -49,6 +49,7 @@ func AuthenticatedClient(options gophercloud.AuthOptions) (*gophercloud.Provider
 	if err != nil {
 		return nil, err
 	}
+
```
Comment (smashwilson) [inline rackspace/client.go]: Whitespace?

File: openstack/client.go [hunk 4]
```diff
@@ -58,6 +58,7 @@ func AuthenticatedClient(options gophercloud.AuthOptions) (*gophercloud.Provider
 	if err != nil {
 		return nil, err
 	}
+
```
Comment (smashwilson) [inline openstack/client.go]: More whitespace?

**Comment by smashwilson** [comment]
> Originally I added a MaxRetries field to the AuthOptions structure to make the number of re-authentication retries variable, but I don't know how helpful that'd be. If there is interest in adding it, it would be just a few lines. Currently it just tries to reauth once.

I like it fine the way it is :grin:. If reauthentication fails, the most likely cause is incorrect credentials (like a cycled API key), which retrying isn't going to help.

It might be worth detecting rate limiting and implementing a retry system with exponential backoff for that -- but that's a different status code and a different issue :wink:

File: openstack/client.go (hunk 1 above)
Comment (jrperritt) [inline openstack/client.go]: :+1:  Good idea

**Comment by smashwilson** [comment]
:metal:
ship it

**Pull request merged**